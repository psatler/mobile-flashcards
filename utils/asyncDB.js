import { AsyncStorage } from 'react-native'
import { data } from './mockData'
const FLASHCARDS_DECKS_KEY = 'Flashcards:decks'

const setInitialData = async () => {
    let obj = data;

    await AsyncStorage.setItem(FLASHCARDS_DECKS_KEY, JSON.stringify(obj))
}


export const getDecks = async () => {
    try {
        let value = await AsyncStorage.getItem(FLASHCARDS_DECKS_KEY);
        if( value !== null){ //we have data
            // console.log('decks', value);
            let parsedObj = JSON.parse(value); //parsing back to object
            return parsedObj;
        } else { //if not, we set initial state
            console.log('setting initial state');
            setInitialData(); //defined above
        }
    } catch (err) {
        console.log('There was an error retrieving decks from local DB: ', err);
    }
}

export const saveDeckTitle = (keyTitle, deckObj) => {
    try {
        return AsyncStorage.mergeItem(FLASHCARDS_DECKS_KEY, JSON.stringify({
            [keyTitle]: deckObj
        }))
        
    } catch (err) {
        console.log(err);
    }
}

export const addCardToDeck = (deckTitle, card) => {
    const key = deckTitle;

    return AsyncStorage.getItem(FLASHCARDS_DECKS_KEY)
        .then( (results) => {
            const allDecks = JSON.parse(results);
            const deck = allDecks[key];
            // console.log('addCardToDeck- BEFORE', deck.questions)
            deck.questions.push(card)
            // console.log('addCardToDeck', deck.questions)
            // console.log('allDecks', allDecks)

            AsyncStorage.setItem(FLASHCARDS_DECKS_KEY, JSON.stringify(allDecks)); //updating 

        })
        // const key = deckTitle.split(' ').join('');

}

// 

// const checkIfDataExists = (results) => {
//     return results === null
//       ? setInitialData()
//       : JSON.parse(results)
//   }

// export const retrieveDataFromAsyncStorage = () => {

//     return AsyncStorage.getItem(FLASHCARDS_DECKS_KEY)
//         .then(checkIfDataExists)
// }



// _setValue = async () => {
//     try {
//       await AsyncStorage.setItem(STORAGE_KEY, 'true');
//     } catch (error) { // log the error
//     }
//   };

//   _loadInitialState = async () => {
//     try {
//       let value = await AsyncStorage.getItem(STORAGE_KEY);
//       if (value === 'true'){
//         this.setState({loaded: 'true'});
//       } else {
//         this.setState({loaded: 'false'});
//         this._setValue();
//       }
//     } catch (error) {
//       this.setState({loaded: 'false'});
//       this._setValue();
//     }
//   };