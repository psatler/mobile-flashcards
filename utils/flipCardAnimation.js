import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  Animated
} from 'react-native';

export default class AnimatedBasic extends Component {

    state = {
        animatedValue: new Animated.Value(0),
        listenerValue: 0,
    }

    componentDidMount(){
        const { animatedValue, listenerValue } = this.state;

        //creating a listener to track values changes
        animatedValue.addListener( ( { value } ) => {
            this.setState({
                listenerValue: value,
            })
        })
    }

    flipCard = () => {
        const { animatedValue, listenerValue } = this.state;

        if(listenerValue >= 90){
            Animated.spring(animatedValue,{
                toValue: 0,
                friction: 8,
                tension: 10
              }).start();
        } 
        else {
            Animated.spring(animatedValue,{
                toValue: 180,
                friction: 8,
                tension: 10
              }).start();
        }
    }

    render() {
        const { animatedValue, listenerValue } = this.state;

        const frontAnimatedStyle = {
          transform: [
            { rotateY: animatedValue.interpolate({
                inputRange: [0, 180],
                outputRange: ['0deg', '180deg'],
              })
            }
          ],
          //adding opacity so the back face won't be mirrored on Android devices
        //https://github.com/facebook/react-native/issues/1973#issuecomment-262059217
          opacity: animatedValue.interpolate({ //adding this so it won't be mirrored on Android devices
                inputRange: [89, 90],
                outputRange: [1, 0]
            }), 
        }
        const backAnimatedStyle = {
          transform: [
            { rotateY: animatedValue.interpolate({
                inputRange: [0, 180],
              //   outputRange: ['180deg', '360deg'],
                outputRange: ['180deg', '0deg']
              }) 
            }
          ],
          opacity: animatedValue.interpolate({  //adding this so it won't be mirrored on Android devices
                inputRange: [89, 90], 
                outputRange: [0, 1] 
            }), 
        }
        
        return (
          <View style={styles.container}>
            <View>
              <Animated.View style={[styles.flipCard, frontAnimatedStyle]}>
                <Text style={styles.flipText}>
                  This text is flipping on the front.
                </Text>
              </Animated.View>
              <Animated.View style={[backAnimatedStyle, styles.flipCard, styles.flipCardBack]}>
                <Text style={styles.flipText}>
                  This text is flipping on the back.
                </Text>
              </Animated.View>
            </View>
            <TouchableOpacity onPress={() => this.flipCard()}>
              <Text>Flip!</Text>
            </TouchableOpacity>
          </View>
        );
      }
    }


const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },
  flipCard: {
    width: 200,
    height: 200,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'blue',
    backfaceVisibility: 'hidden', //it means when the card rotates out of the way, it won't be visible. When it rotates back, it'll be visible again. Not working on Android though, so it was added an workaround for this above
  },
  flipCardBack: {
    backgroundColor: "red",
    position: "absolute",
    top: 0,
  },
  flipText: {
    width: 90,
    fontSize: 20,
    color: 'white',
    fontWeight: 'bold',
  }
});