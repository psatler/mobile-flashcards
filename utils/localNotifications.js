// import React from 'react'
import { AsyncStorage } from 'react-native'
import { Notifications, Permissions } from 'expo'

const FLASHCARDS_NOTIFICATIONS_KEY = 'Flashcards:localNotifications'


const createNotification = () => {
    return {
        title: 'Do a quiz today!',
        body: 'ðŸ‘‹ Do not forget to study your decks!',
        ios: {
            sound: true,
        },
        android: {
            sound: true,
            priority: 'high',
            sticky: false,
            vibrate: true,
        }
    }
}

export const setLocalNotification = () => {
    //checking if our local notification has already been set
    AsyncStorage.getItem(FLASHCARDS_NOTIFICATIONS_KEY) 
        .then(JSON.parse) //parsing the returned value
        .then( (data) => {
            if(data === null){ //if we haven't any notification scheduled
                Permissions.askAsync(Permissions.NOTIFICATIONS)
                    .then( ({ status }) => {
                        if( status === 'granted' ){
                            Notifications.cancelAllScheduledNotificationsAsync();

                            let tomorrow = new Date();
                            tomorrow.setDate(tomorrow.getDate() + 1) //tomorrow 
                            tomorrow.setHours(10); //scheduling for 10 AM
                            tomorrow.setMinutes(0);

                            Notifications.scheduleLocalNotificationAsync(
                                createNotification(),
                                {
                                    time: tomorrow,
                                    repeat: 'day'
                                }
                            )

                            AsyncStorage.setItem(FLASHCARDS_NOTIFICATIONS_KEY, JSON.stringify(true))
                        }
                    })

            }
        })

}

export const clearLocalNotification = () => {
    return AsyncStorage.removeItem(FLASHCARDS_NOTIFICATIONS_KEY)
      .then(Notifications.cancelAllScheduledNotificationsAsync)
}