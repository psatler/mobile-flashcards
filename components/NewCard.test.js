import React from 'react'
import ConnectedNewCard, { NewCard } from './NewCard'
import { data } from '../utils/mockData'

import renderer from 'react-test-renderer';

//####################################### ENZYME STUFF
import { configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
configure({ adapter: new Adapter() })
import { shallow } from 'enzyme'

//####################################### REDUX MOCK STORE
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
const mockStore = configureMockStore([ thunk ]);
// const initialState = data['React'];
const initialState = { }
const storeMock = mockStore(initialState);


describe('[Component] NewCard', () => {
    const mockData = data['JavaScript'];
    const navigation = {
        goBack: jest.fn(),
        state: {
            params: {
                title: mockData.title,
            }
        }
    };

    afterEach( () => {
        jest.resetAllMocks();
    })

    it('shallow renders correctly', () => {
        const wrapper = shallow(<NewCard />);
        expect(wrapper).toMatchSnapshot();
    });

    it('shows the alert message when user press submit button with blanks fields', () => {
        //####### mocking the alert method
        global.alert = jest.fn(); 
        const wrapper = shallow(<NewCard navigation={navigation}  />);
        wrapper.find('TouchableOpacity').first().props().onPress() //first touchableOpacity
        expect(global.alert).toHaveBeenCalledTimes(1);
    });

    it('adds card to deck when user press submit button', () => {
        const dispatch = jest.fn();
        const wrapper = shallow(<NewCard navigation={navigation} dispatch={dispatch} /> );
        wrapper.setState({
            questionInput: 'Testing question',
            answerInput: 'Testing answer',
        })
        wrapper.find('TouchableOpacity').first().props().onPress() //first touchableOpacity
        expect(navigation.goBack).toHaveBeenCalledTimes(1);
        // expect(wrapper.state.questionInput).toEqual('');
    });

    it('calls handleQuestionTextChange and handleAnswerTextChange', () => {
        const wrapper = shallow(<NewCard navigation={navigation}  />)
        // console.log('test', wrapper.state())
        wrapper.find('TextInput').first().props().onChangeText() //question 
        wrapper.find('TextInput').last().props().onChangeText() //answer 
    });

    xit('calls onSubmitEditing and goes to the next input field', () => {
        const wrapper = shallow(<NewCard navigation={navigation}  />)
        // console.log('test', wrapper.state())
        wrapper.find('TextInput').first().props().onSubmitEditing() //question 
        // wrapper.find('TextInput').last().props().onChangeText() //answer 
    });

    xit('adds card to deck when user press submit button', () => {
        
        const wrapper = shallow(<ConnectedNewCard navigation={navigation}  />, 
            { context: { store: storeMock } },
        );
        wrapper.setState({
            questionInput: 'Testing question',
            answerInput: 'Testing answer',
        })
        wrapper.dive().find('TouchableOpacity').first().props().onPress() //first touchableOpacity
        expect(navigation.goBack).toHaveBeenCalledTimes(1);
    });

    // xit('calls handleQuestionTextChange and handleAnswerTextChange', () => {
    //     const wrapper = shallow(<NewCard navigation={navigation}  />)
    //     console.log('test', wrapper.state())
    //     const input = wrapper.find('TextInput').first();
    //     input.simulate('change', { target: { value: 'Hello' } })
    //     wrapper.find('TextInput').first().props().onChangeText() //question 
    // });


    // it('calls getDecks method inside componentDidMount', () => {
    //     const spy = jest.spyOn(NewCard.prototype, 'submitDeck');
    //     const wrapper = shallow(<NewCard  />);
    //     wrapper.instance().submitDeck();
    //     expect(spy).toHaveBeenCalled();
    // })

    
});