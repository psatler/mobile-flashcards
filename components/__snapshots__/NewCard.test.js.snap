// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] NewCard shallow renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NewCard />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "behavior": "padding",
      "children": Array [
        <TextInput
          allowFontScaling={true}
          blurOnSubmit={false}
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Insert a question"
          style={
            Object {
              "borderColor": "#00A4D3",
              "borderRadius": 16,
              "borderWidth": 1,
              "height": 50,
              "marginTop": 20,
              "paddingLeft": 15,
              "width": 720,
            }
          }
          value=""
        />,
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          placeholder="Insert an answer"
          style={
            Object {
              "borderColor": "#00A4D3",
              "borderRadius": 16,
              "borderWidth": 1,
              "height": 50,
              "marginTop": 20,
              "paddingLeft": 15,
              "width": 720,
            }
          }
          value=""
        />,
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
          style={
            Object {
              "backgroundColor": "#18263D",
              "borderRadius": 16,
              "marginTop": 25,
              "padding": 10,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#00A4D3",
                "fontSize": 15,
                "padding": 5,
              }
            }
          >
             Submit 
          </Text>
        </TouchableOpacity>,
      ],
      "enabled": true,
      "keyboardVerticalOffset": 0,
      "style": Object {
        "alignItems": "center",
        "flex": 1,
        "padding": 10,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "allowFontScaling": true,
          "blurOnSubmit": false,
          "onChangeText": [Function],
          "onSubmitEditing": [Function],
          "placeholder": "Insert a question",
          "style": Object {
            "borderColor": "#00A4D3",
            "borderRadius": 16,
            "borderWidth": 1,
            "height": 50,
            "marginTop": 20,
            "paddingLeft": 15,
            "width": 720,
          },
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "allowFontScaling": true,
          "onChangeText": [Function],
          "placeholder": "Insert an answer",
          "style": Object {
            "borderColor": "#00A4D3",
            "borderRadius": 16,
            "borderWidth": 1,
            "height": 50,
            "marginTop": 20,
            "paddingLeft": 15,
            "width": 720,
          },
          "value": "",
        },
        "ref": [Function],
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "activeOpacity": 0.2,
          "children": <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#00A4D3",
                "fontSize": 15,
                "padding": 5,
              }
            }
          >
             Submit 
          </Text>,
          "onPress": [Function],
          "style": Object {
            "backgroundColor": "#18263D",
            "borderRadius": 16,
            "marginTop": 25,
            "padding": 10,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "accessible": true,
            "allowFontScaling": true,
            "children": " Submit ",
            "ellipsizeMode": "tail",
            "style": Object {
              "color": "#00A4D3",
              "fontSize": 15,
              "padding": 5,
            },
          },
          "ref": null,
          "rendered": " Submit ",
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "behavior": "padding",
        "children": Array [
          <TextInput
            allowFontScaling={true}
            blurOnSubmit={false}
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            placeholder="Insert a question"
            style={
              Object {
                "borderColor": "#00A4D3",
                "borderRadius": 16,
                "borderWidth": 1,
                "height": 50,
                "marginTop": 20,
                "paddingLeft": 15,
                "width": 720,
              }
            }
            value=""
          />,
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Insert an answer"
            style={
              Object {
                "borderColor": "#00A4D3",
                "borderRadius": 16,
                "borderWidth": 1,
                "height": 50,
                "marginTop": 20,
                "paddingLeft": 15,
                "width": 720,
              }
            }
            value=""
          />,
          <TouchableOpacity
            activeOpacity={0.2}
            onPress={[Function]}
            style={
              Object {
                "backgroundColor": "#18263D",
                "borderRadius": 16,
                "marginTop": 25,
                "padding": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#00A4D3",
                  "fontSize": 15,
                  "padding": 5,
                }
              }
            >
               Submit 
            </Text>
          </TouchableOpacity>,
        ],
        "enabled": true,
        "keyboardVerticalOffset": 0,
        "style": Object {
          "alignItems": "center",
          "flex": 1,
          "padding": 10,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "allowFontScaling": true,
            "blurOnSubmit": false,
            "onChangeText": [Function],
            "onSubmitEditing": [Function],
            "placeholder": "Insert a question",
            "style": Object {
              "borderColor": "#00A4D3",
              "borderRadius": 16,
              "borderWidth": 1,
              "height": 50,
              "marginTop": 20,
              "paddingLeft": 15,
              "width": 720,
            },
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "allowFontScaling": true,
            "onChangeText": [Function],
            "placeholder": "Insert an answer",
            "style": Object {
              "borderColor": "#00A4D3",
              "borderRadius": 16,
              "borderWidth": 1,
              "height": 50,
              "marginTop": 20,
              "paddingLeft": 15,
              "width": 720,
            },
            "value": "",
          },
          "ref": [Function],
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "activeOpacity": 0.2,
            "children": <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#00A4D3",
                  "fontSize": 15,
                  "padding": 5,
                }
              }
            >
               Submit 
            </Text>,
            "onPress": [Function],
            "style": Object {
              "backgroundColor": "#18263D",
              "borderRadius": 16,
              "marginTop": 25,
              "padding": 10,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "accessible": true,
              "allowFontScaling": true,
              "children": " Submit ",
              "ellipsizeMode": "tail",
              "style": Object {
                "color": "#00A4D3",
                "fontSize": 15,
                "padding": 5,
              },
            },
            "ref": null,
            "rendered": " Submit ",
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
